internal/multiparty/manager.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/hyperledger/firefly/pkg/core"
+)
+
+type Manager interface {
+	// ConfigureContract initializes the subscription to the FireFly contract
+	// - Checks the provided contract info against the plugin's configuration, and updates it as needed
+	// - Initializes the contract info for performing BatchPin transactions, and initializes subscriptions for BatchPin events
+	ConfigureContract(ctx context.Context, contracts *core.FireFlyContracts) (err error)
+
+	// TerminateContract marks the given event as the last one to be parsed on the current FireFly contract
+	// - Validates that the event came from the currently active FireFly contract
+	// - Re-initializes the plugin against the next configured FireFly contract
+	// - Updates the provided contract info to record the point of termination and the newly active contract
+	TerminateContract(ctx context.Context, contracts *core.FireFlyContracts, termination *blockchain.Event) (err error)
+
+	// GetNetworkVersion returns the network version of the active FireFly contract
+	GetNetworkVersion() int
+
+	// SubmitBatchPin sequences a batch of message globally to all viewers of a given ledger
+	SubmitBatchPin(ctx context.Context, nsOpID string, signingKey string, batch *blockchain.BatchPin) error
+
+	// SubmitNetworkAction writes a special "BatchPin" event which signals the plugin to take an action
+	SubmitNetworkAction(ctx context.Context, nsOpID string, signingKey string, action core.NetworkActionType) error
+
+	core.Named
+}
+
+type Contract struct {
+	Location   *fftypes.JSONAny
+	FirstEvent string
+}
+
+type multipartyManager struct {
+	ctx            context.Context
+	blockchain     blockchain.Plugin
+	contracts      []Contract
+	namespace      string
+	activeContract struct {
+		location       *fftypes.JSONAny
+		mux            sync.Mutex
+		firstEvent     string
+		networkVersion int
+		subscription   string
+	}
+}
+
+func NewMultipartyManager(ctx context.Context, ns string, contracts []Contract, bi blockchain.Plugin) Manager {
+	return &multipartyManager{
+		ctx:        ctx,
+		namespace:  ns,
+		contracts:  contracts,
+		blockchain: bi,
+	}
+}
+
+func (mm *multipartyManager) ConfigureContract(ctx context.Context, contracts *core.FireFlyContracts) (err error) {
+	log.L(ctx).Infof("Resolving FireFly contract at index %d", contracts.Active.Index)
+	location, firstEvent, err := mm.resolveFireFlyContract(ctx, contracts.Active.Index)
+	if err != nil {
+		return err
+	}
+	version, err := mm.blockchain.GetNetworkVersion(ctx, location)
+	if err != nil {
+		return err
+	}
+
+	subID, err := mm.blockchain.AddFireflySubscription(ctx, mm.namespace, location, firstEvent)
+	if err == nil {
+		mm.activeContract.mux.Lock()
+		mm.activeContract.location = location
+		mm.activeContract.firstEvent = firstEvent
+		mm.activeContract.networkVersion = version
+		mm.activeContract.subscription = subID
+		mm.activeContract.mux.Unlock()
+		contracts.Active.Info = fftypes.JSONObject{
```

Comment:
Originally `Info` was a `JSONObject` because it was expected to be different per blockchain. It looks like it now has a predictable blockchain-agnostic structure, so it could be a well-defined `struct`.

Reply:
Or maybe these fields are just directly on `Active` and we don't need a nested `Info` at all...